kernel, logging: Pass Logger instances to kernel objects

ryanofsky opened the pull request:
Pass `Logger` instances to `BlockManager`, `CCoinsViewDB`, `CDBWrapper`, `Chainstate`, `ChainstateManager`, `CoinsViews`, and `CTxMemPool` classes so libbitcoinkernel applications and tests have the option to control where log output goes instead of having all output sent to the global logger.

This PR is an alternative to #30338. It is more limited because it only changes kernel code while leaving other parts of the codebase alone. It also takes the opportunity to migrate legacy LogPrint / LogPrintf calls to the new log macros introduced in #28318.

---

<!-- begin based-on -->
**This is based on #34778 + #29256 + #33847.** The non-base commits are:

- [`5fa829c394d` refactor: Pass Logger instances to kernel objects](https://github.com/bitcoin/bitcoin/pull/30342/commits/5fa829c394d414c4e3926fcd93451a9981d66694)
- [`52caf3c63fd` logging: Add LOG_REQUIRE_CONTEXT option](https://github.com/bitcoin/bitcoin/pull/30342/commits/52caf3c63fd2150ecd5959d32a52fbb73a84024c)
- [`299a6c593bf` refactor: Log kernel output to local log instances](https://github.com/bitcoin/bitcoin/pull/30342/commits/299a6c593bf1de9454c92b2d8d1dfdcf449d092f)
- [`846df5f6f47` kernel: Drop global Logger instance](https://github.com/bitcoin/bitcoin/pull/30342/commits/846df5f6f474f727f13075ede6e4ce385e5e0bea)<!-- end -->

ryanofsky committed:
log test: Add test for all accepted logging arguments

Add new logging test to call macros with all allowed combinations of macro
arguments.

The new test replaces a less comprehensive test that doesn't cover log
statements without format arguments. It's also moved to the top of the test
suite because it's a good illustration of what typical log prints look like,
what logging calls are allowed and disallowed, and what the resulting log
output looks like.

ryanofsky committed:
doc: Add documentation about log levels and macros

ryanofsky committed:
log refactor: Allow log macros to accept context arguments

Allow LogDebug(), LogTrace(), LogInfo(), LogWarning(), and LogError() macros to
accept context arguments to provide more information in log messages and more
control over logging to callers.

This functionality is used in followup PRs:

- https://github.com/bitcoin/bitcoin/pull/30342 - To let libbitcoinkernel send
  output to specfic `BCLog::Logger` instances instead of a global instance, so
  output can be disambiguated and applications can have more control over
  logging.

- https://github.com/bitcoin/bitcoin/pull/30343 - To replace custom
  `WalletLogPrintf` calls with standard logging calls that automatically include
  wallet names and don't log everything at info level.

This commit does not change behavior of current log prints or require them to
be updated. It includes tests and documentation covering the new functionality.

Co-Authored-By: Hodlinator <172445034+hodlinator@users.noreply.github.com>
Co-Authored-By: stickies-v <stickies-v@protonmail.com>

ryanofsky committed:
refactor: Pass Logger instances to kernel objects

Pass Logger instances to BlockManager, CCoinsViewDB, CDBWrapper,
ChainstateManager, and CoinsViews instances so libbitcoinkernel applications
and test code have the option to control where log output goes instead of
having all output sent to the global logger.

This commit just passes the logger objects without using them. The next commit
updates log print statements to use the new objects.

ryanofsky committed:
refactor: Log kernel output to local log instances

This is a mechanical change updating kernel code that currently uses the global
log instance to log to local instances instead.

TheCharlatan commented:
Concept ACK.

ryanofsky commented:
Updated 45423355735770f01d2bac738cc8b0b41415e921 -> db1b9f7696af80036de739408cd9eae5d06481ec ([`pr/gklog.1`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.1) -> [`pr/gklog.2`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.1..pr/gklog.2)) moving code to an earlier commit to fix a "test-each-commit" test failure https://github.com/bitcoin/bitcoin/actions/runs/9684398780/job/26722073921?pr=30342, and cleaning up code to fix a clang-tidy warning https://cirrus-ci.com/task/5893151045451776

ryanofsky committed:
log refactor: Ensure categories are not logged at info and higher levels

Previously this used to be possible through the LogPrintLevel call but now that
call is removed, this change is just an internal refactoring and has no visible
effect except in tests.

ryanofsky committed:
kernel: Drop global Logger instance

Change LogInstance() function to no longer allocate (and leak) a BCLog::Logger
instance. Instead allow kernel applications to initialize their own logging
instances that can be returned by LogInstance().

The LogInstance() function is not actually used for the majority of logging in
kernel code. Most kernel log output goes directly to BCLog::Logger instances
specified when kernel objects like ChainstateManager and CTxMemPool are
constructed, which allows applications to create multiple Logger instances and
control which log output is sent to them.

The only kernel code that does rely on LogInstance() is certain low level code in
the util library, like the RNG seeder, that is not passed a specific instance
and needs to rely on the global LogInstance() function.

Other code outside the kernel library uses LogInstance() heavily, and may
continue to do so. bitcoind and test code now just create a log instance before
the first LogInstance() call, which it returns, so all behavior is the same as
before.

ryanofsky commented:
Rebased db1b9f7696af80036de739408cd9eae5d06481ec -> 42d0e06d1a4343c2bf3257f26cf18b79072c6b0d ([`pr/gklog.2`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.2) -> [`pr/gklog.3`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.2-rebase..pr/gklog.3)) to fix conflict with #30141. Also added a commit to let kernel applications direct low-level util output to custom log instances.
Updated 42d0e06d1a4343c2bf3257f26cf18b79072c6b0d -> fcaed3faf0f12b3b322e65df1e81d455bbee7db5 ([`pr/gklog.3`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.3) -> [`pr/gklog.4`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.3..pr/gklog.4)) to fix CI errors
Updated fcaed3faf0f12b3b322e65df1e81d455bbee7db5 -> f005677f762754ae09417c2a418e27d28ef3c7e1 ([`pr/gklog.4`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.4) -> [`pr/gklog.5`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.4..pr/gklog.5)) to fix more CI errors
Rebased f005677f762754ae09417c2a418e27d28ef3c7e1 -> dcd055ecf9c20b84930c2ce45d922427cf0112ea ([`pr/gklog.5`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.5) -> [`pr/gklog.6`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.5-rebase..pr/gklog.6)) due to conflicts with #30425 and #30407
Rebased dcd055ecf9c20b84930c2ce45d922427cf0112ea -> 53d7933727ebe0670dceb88ec9f1020bd7b0b280 ([`pr/gklog.6`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.6) -> [`pr/gklog.7`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.6-rebase..pr/gklog.7)) due to conflict with #28052
Rebased 53d7933727ebe0670dceb88ec9f1020bd7b0b280 -> 10c275342dca365a6678da5fc016ca9d46d9dfca ([`pr/gklog.7`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.7) -> [`pr/gklog.8`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.7-rebase..pr/gklog.8)) due to conflict with #30485
Rebased 10c275342dca365a6678da5fc016ca9d46d9dfca -> 493dd26b75d0e35422b3230be3c72a0f96eb5e8a ([`pr/gklog.8`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.8) -> [`pr/gklog.9`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.8-rebase..pr/gklog.9)) due to various conflicts
Rebased 493dd26b75d0e35422b3230be3c72a0f96eb5e8a -> e971028fa5e570acf585073494acd65f4831ab1c ([`pr/gklog.9`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.9) -> [`pr/gklog.10`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.9-rebase..pr/gklog.10)) due to conflicts with #31393, #31490, #30965
<!-- begin push-11 -->
Rebased e971028fa5e570acf585073494acd65f4831ab1c -> 282db88a9b1789f0d78928c87c7204402ac11ac3 ([`pr/gklog.10`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.10) -> [`pr/gklog.11`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.11), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.10-rebase..pr/gklog.11))<!-- end -->
<!-- begin push-12 -->
Rebased 282db88a9b1789f0d78928c87c7204402ac11ac3 -> ec6d771d3b34ec979e373f025c4f064ae44df860 ([`pr/gklog.11`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.11) -> [`pr/gklog.12`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.12), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.11-rebase..pr/gklog.12))<!-- end -->
<!-- begin push-13 -->
Rebased ec6d771d3b34ec979e373f025c4f064ae44df860 -> b1b72969c9764636721a7518a2f4453d0eb3845d ([`pr/gklog.12`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.12) -> [`pr/gklog.13`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.13), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.12-rebase..pr/gklog.13))<!-- end -->
<!-- begin push-14 -->
Rebased b1b72969c9764636721a7518a2f4453d0eb3845d -> d0e2944bb18e44c599312b980900c324ae698464 ([`pr/gklog.13`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.13) -> [`pr/gklog.14`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.14), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.13-rebase..pr/gklog.14))<!-- end --> on updated base PR to fix CI errors (https://github.com/bitcoin/bitcoin/actions/runs/18514131135?pr=30342)
<!-- begin push-15 -->
Rebased d0e2944bb18e44c599312b980900c324ae698464 -> 8482a49ab89012c3cd99acd43e1f698311426c3a ([`pr/gklog.14`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.14) -> [`pr/gklog.15`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.15), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.14-rebase..pr/gklog.15))<!-- end -->
<!-- begin push-16 -->
Rebased 8482a49ab89012c3cd99acd43e1f698311426c3a -> baaaa41b93cba064d12d641899968ee897eb0671 ([`pr/gklog.15`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.15) -> [`pr/gklog.16`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.16), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.15-rebase..pr/gklog.16))<!-- end --> to fix fuzz test errors https://github.com/bitcoin/bitcoin/actions/runs/19249628369/job/55031630303?pr=30342, also making updates to bitcoinkernel.h documentation
<!-- begin push-17 -->
Rebased baaaa41b93cba064d12d641899968ee897eb0671 -> b9dc75a2b5d6bfadce65cd2837a3042c888a8afd ([`pr/gklog.16`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.16) -> [`pr/gklog.17`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.17), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.16-rebase..pr/gklog.17))<!-- end -->
<!-- begin push-18 -->
Rebased b9dc75a2b5d6bfadce65cd2837a3042c888a8afd -> 97be3ef0b6239a9e674be3d8c70f99ec5fe5b8eb ([`pr/gklog.17`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.17) -> [`pr/gklog.18`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.18), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.17-rebase..pr/gklog.18))<!-- end -->
<!-- begin push-19 -->
Rebased 97be3ef0b6239a9e674be3d8c70f99ec5fe5b8eb -> db29e8fb43fa8011eb5c68f376ea74b4f087deda ([`pr/gklog.18`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.18) -> [`pr/gklog.19`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.19), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.18-rebase..pr/gklog.19))<!-- end --> on top of #29256 pr/bclog.34 and #33847 pr/klog.4
<!-- begin push-20 -->
Updated db29e8fb43fa8011eb5c68f376ea74b4f087deda -> d6c23d4219b30dc8e045f4c6b90a7ad493faf383 ([`pr/gklog.19`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.19) -> [`pr/gklog.20`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.20), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.19..pr/gklog.20))<!-- end --> to fix kernel test MSAN error https://github.com/bitcoin/bitcoin/actions/runs/21115418786/job/60720235987?pr=30342
<!-- begin push-21 -->
Rebased d6c23d4219b30dc8e045f4c6b90a7ad493faf383 -> 75e1e958f5e8a27491723aa88c69bf3cd8de7509 ([`pr/gklog.20`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.20) -> [`pr/gklog.21`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.21), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.20-rebase..pr/gklog.21))<!-- end --> on top of #29256 pr/bclog.34 and #33847 pr/klog.5
<!-- begin push-22 -->
Rebased 75e1e958f5e8a27491723aa88c69bf3cd8de7509 -> cc0cfe20c8ac56ffe3a47d9ee3f109fdf0c09c36 ([`pr/gklog.21`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.21) -> [`pr/gklog.22`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.22), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.21-rebase..pr/gklog.22))<!-- end --> due to conflicts with #34253 and #33680
<!-- begin push-23 -->
Updated cc0cfe20c8ac56ffe3a47d9ee3f109fdf0c09c36 -> c2e3362b0050483e5a95c9aaccb9d3afe6328c1e ([`pr/gklog.22`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.22) -> [`pr/gklog.23`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.23), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.22..pr/gklog.23))<!-- end --> to fix CI errors https://github.com/bitcoin/bitcoin/actions/runs/21643951237
<!-- begin push-24 -->
Updated c2e3362b0050483e5a95c9aaccb9d3afe6328c1e -> 7b78335f66b08d039f20dc31fb326883ab5ad3c8 ([`pr/gklog.23`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.23) -> [`pr/gklog.24`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.24), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.23..pr/gklog.24))<!-- end --> to fix CI errors https://github.com/bitcoin/bitcoin/actions/runs/21645374775
<!-- begin push-25 -->
Rebased 7b78335f66b08d039f20dc31fb326883ab5ad3c8 -> aa4d26979f0b1c24adeeec489b36180c87ca484c ([`pr/gklog.24`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.24) -> [`pr/gklog.25`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.25), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.24-rebase..pr/gklog.25))<!-- end --> on top of #29256 pr/bclog.36 and #33847 pr/klog.5
<!-- begin push-26 -->
Updated aa4d26979f0b1c24adeeec489b36180c87ca484c -> 5126eaf1e369f1579d2307125aa8383786f9b3f8 ([`pr/gklog.25`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.25) -> [`pr/gklog.26`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.26), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.25..pr/gklog.26))<!-- end --> to fix IWYU and each-commit errors https://github.com/bitcoin/bitcoin/actions/runs/21830259022/job/62995439943?pr=30342
<!-- begin push-27 -->
Updated 5126eaf1e369f1579d2307125aa8383786f9b3f8 -> eeb84002a84226614e5d3acb9013c15ba5ce71ff ([`pr/gklog.26`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.26) -> [`pr/gklog.27`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.27), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.26..pr/gklog.27))<!-- end --> to fix IWYU errors https://github.com/bitcoin/bitcoin/actions/runs/21833389551/job/62998114427?pr=30342
<!-- begin push-28 -->
Rebased eeb84002a84226614e5d3acb9013c15ba5ce71ff -> d137676ffd09da75f0418f7787fab8e249824e6f ([`pr/gklog.27`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.27) -> [`pr/gklog.28`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.28), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.27-rebase..pr/gklog.28))<!-- end --> to fix conflicts with #32950, #34165, #33512 on top of #29256 pr/bclog.37
<!-- begin push-29 -->
Updated d137676ffd09da75f0418f7787fab8e249824e6f -> ea7842f5b68c2986982c31f2482d47d2bc89c382 ([`pr/gklog.28`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.28) -> [`pr/gklog.29`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.29), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.28..pr/gklog.29))<!-- end --> to fix silent conflicts with threadpool tests #34562 causing windows build errors and fuzz test failures in ci https://github.com/bitcoin/bitcoin/actions/runs/22559603524/job/65396727322?pr=30342
<!-- begin push-30 -->
Updated ea7842f5b68c2986982c31f2482d47d2bc89c382 -> b6b95a62447c589a0078474fb8f6bc4b2e7e8a51 ([`pr/gklog.29`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.29) -> [`pr/gklog.30`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.30), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.29..pr/gklog.30))<!-- end --> to fix threadpool fuzz test cleanup https://github.com/bitcoin/bitcoin/actions/runs/22584916974/job/65427016231?pr=30342
<!-- begin push-31 -->
Rebased b6b95a62447c589a0078474fb8f6bc4b2e7e8a51 -> f048e807c6b4327474466c749fe48cb43dd651a0 ([`pr/gklog.30`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.30) -> [`pr/gklog.31`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.31), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.30-rebase..pr/gklog.31))<!-- end --> on top of #29256 pr/bclog.39
<!-- begin push-32 -->
Updated f048e807c6b4327474466c749fe48cb43dd651a0 -> 71e2a726cb812f40869c8e9dd09257ffbdb4ab05 ([`pr/gklog.31`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.31) -> [`pr/gklog.32`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.32), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.31..pr/gklog.32))<!-- end --> to fix IWYU error in `index/base.cpp` https://github.com/bitcoin/bitcoin/actions/runs/22861803182/job/66317637780?pr=30342. Not really sure the IWYU error makes sense but implemented the suggestion
<!-- begin push-33 -->
Rebased 71e2a726cb812f40869c8e9dd09257ffbdb4ab05 -> 74c694e7a55ed3d77bd70eef3b66fed7c5edb222 ([`pr/gklog.32`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.32) -> [`pr/gklog.33`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.33), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.32-rebase..pr/gklog.33))<!-- end --> on top of #29256 pr/bclog.40
<!-- begin push-34 -->
Squashed 74c694e7a55ed3d77bd70eef3b66fed7c5edb222 -> f3a87a203f8c6916786916200368701132f419d1 ([`pr/gklog.33`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.33) -> [`pr/gklog.34`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.34), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.33..pr/gklog.34))<!-- end --> to fix compile error in internal commit https://github.com/bitcoin/bitcoin/actions/runs/23868674296/job/69594579042?pr=30342
<!-- begin push-35 -->
Rebased f3a87a203f8c6916786916200368701132f419d1 -> 41fb7ccf129bee40b19e5968a7ee8b39c6893544 ([`pr/gklog.34`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.34) -> [`pr/gklog.35`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.35), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.34-rebase..pr/gklog.35))<!-- end --> on top of #29256 pr/bclog.41
<!-- begin push-36 -->
Updated 41fb7ccf129bee40b19e5968a7ee8b39c6893544 -> 6e7439bf9c5d0db4332ad71949ffad5cf0c91329 ([`pr/gklog.35`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.35) -> [`pr/gklog.36`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.36), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.35..pr/gklog.36))<!-- end --> to fix ci error https://github.com/bitcoin/bitcoin/actions/runs/25750656813/job/75625762044 to fix test errors in test_kernel from a missing logger in  btck_check_block_context_free in certain test orders
<!-- begin push-37 -->
Rebased 6e7439bf9c5d0db4332ad71949ffad5cf0c91329 -> eeaa0f75f4fdb23b26cb0218c5b85d0c6203f978 ([`pr/gklog.36`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.36) -> [`pr/gklog.37`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.37), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.36-rebase..pr/gklog.37))<!-- end --> on top of #29256 pr/bclog.42
<!-- begin push-38 -->
Updated eeaa0f75f4fdb23b26cb0218c5b85d0c6203f978 -> 8ae06e76b2daacec338a590c5c9cf5216ca0d7df ([`pr/gklog.37`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.37) -> [`pr/gklog.38`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.38), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.37..pr/gklog.38))<!-- end --> to fix CI: iwyu in sigcache.h, duplicate include in bitcoind.cpp, fuzzer null g_setup in dbwrapper.cpp https://github.com/bitcoin/bitcoin/actions/runs/26643336975/job/78525287457
<!-- begin push-39 -->
Rebased 8ae06e76b2daacec338a590c5c9cf5216ca0d7df -> d70eae19a8b6024723f33f29907a3afc1a0a9ca1 ([`pr/gklog.38`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.38) -> [`pr/gklog.39`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.39), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.38-rebase..pr/gklog.39))<!-- end --> on top of #29256 pr/bclog.42 and #33847 pr/klog.5 due to conflicts with #34806, #34865, #34866, and #34887
<!-- begin push-40 -->
Rebased d70eae19a8b6024723f33f29907a3afc1a0a9ca1 -> b3a16a1ce6af8d32f628c9e95d1a2520e6fc34f0 ([`pr/gklog.39`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.39) -> [`pr/gklog.40`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.40), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.39-rebase..pr/gklog.40))<!-- end --> on top of #29256 pr/bclog.42 and #33847 pr/klog.5 due to conflicts with #35465 and #35521
<!-- begin push-41 -->
Squashed b3a16a1ce6af8d32f628c9e95d1a2520e6fc34f0 -> 459562de2121f94d55d0a6bfe86ed11a6c2b59af ([`pr/gklog.40`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.40) -> [`pr/gklog.41`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.41), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.40..pr/gklog.41))<!-- end --> to fix test ancestor commits: pass m_logger to CCoinsViewDB in coins_db_leveldb_layout test https://github.com/bitcoin/bitcoin/actions/runs/28108478624/job/83229088477; and to fix fuzzer crash: stop dbwrapper_concurrent_reads thread pool before logger is destroyed https://github.com/bitcoin/bitcoin/actions/runs/28108478624/job/83229088738
<!-- begin push-42 -->
Rebased 459562de212c27108832bb4d2822d7eda9a8b98e -> 2e611706fce63745a9bf08db4e4ca1631279e00e ([`pr/gklog.41`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.41) -> [`pr/gklog.42`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.42), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.41-rebase..pr/gklog.42))<!-- end --> due to conflicts with #35568 and #35295
<!-- begin push-43 -->
Updated 2e611706fce63745a9bf08db4e4ca1631279e00e -> 372179265729d0b0418bb9ec4dd6b05e812ebc0e ([`pr/gklog.42`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.42) -> [`pr/gklog.43`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.43), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.42..pr/gklog.43))<!-- end --> to fix CI: duplicate BOOST_FIXTURE_TEST_SUITE in coinsviewoverlay_tests.cpp broke cmake configure https://github.com/bitcoin/bitcoin/actions/runs/29337611210/job/87100747364; and missing log context in CoinsViews::InitCache triggered LOG_REQUIRE_CONTEXT assertion in MSVC fuzz build https://github.com/bitcoin/bitcoin/actions/runs/29337611210/job/87100747283
<!-- begin push-44 -->
Updated 372179265729d0b0418bb9ec4dd6b05e812ebc0e -> 532547978dac2d144fb92019b8b2007abb0e95ed ([`pr/gklog.43`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.43) -> [`pr/gklog.44`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.44), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.43..pr/gklog.44))<!-- end --> to fix CI compile errors caused by silent conflicts with #35295 https://github.com/bitcoin/bitcoin/actions/runs/29343064668/job/87119598397
<!-- begin push-45 -->
Updated 532547978dac2d144fb92019b8b2007abb0e95ed -> 9884477bcdc8abce1b640c66615c0ae44e0f9400 ([`pr/gklog.44`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.44) -> [`pr/gklog.45`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.45), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.44..pr/gklog.45))<!-- end --> to fix CI: stop thread pool in coins_view fuzz targets before logger destroyed (https://github.com/bitcoin/bitcoin/actions/runs/29347701435/job/87135632047); store util::log::Context by value in CBitcoinLevelDBLogger to fix dangling ref SIGSEGV in feature_init.py (https://github.com/bitcoin/bitcoin/actions/runs/29347701435/job/87135631969)
<!-- begin push-46 -->
Updated 9884477bcdc8abce1b640c66615c0ae44e0f9400 -> aa9bf0282d833956a41758a474282a1f765530a5 ([`pr/gklog.45`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.45) -> [`pr/gklog.46`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.46), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.45..pr/gklog.46))<!-- end --> to fix CI: thread-safety analysis missing EXCLUSIVE_LOCKS_REQUIRED(!g_thread_pool_mutex) annotation in cleanup_coins_view() https://github.com/bitcoin/bitcoin/actions/runs/29352995762/job/87153614524
<!-- begin push-47 -->
Updated aa9bf0282d833956a41758a474282a1f765530a5 -> d62ba526841ccb30912fe7ca7eac6265fd1cba43 ([`pr/gklog.46`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.46) -> [`pr/gklog.47`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.47), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.46..pr/gklog.47))<!-- end --> to fix fuzzer crash: stop coinscache_sim thread pool before logger is destroyed https://github.com/bitcoin/bitcoin/actions/runs/29355819465/job/87163129249
<!-- begin push-48 -->
Rebased d62ba526841ccb30912fe7ca7eac6265fd1cba43 -> 9cf188afa4742a181f5ac6a892a47e8be66ddc42 ([`pr/gklog.47`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.47) -> [`pr/gklog.48`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.48), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.47-rebase..pr/gklog.48))<!-- end --> due to conflict with #35723
<!-- begin push-49 -->
Updated 9cf188afa4742a181f5ac6a892a47e8be66ddc42 -> 846df5f6f474f727f13075ede6e4ce385e5e0bea ([`pr/gklog.48`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.48) -> [`pr/gklog.49`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.49), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.48..pr/gklog.49))<!-- end --> to fix CI: undeclared g_thread_pool_mutex in cleanup functions (silent conflict with #35723 in push-48 rebase) https://github.com/bitcoin/bitcoin/actions/runs/29372633900/job/87219192227
<!-- begin push-50 -->
Rebased 846df5f6f474f727f13075ede6e4ce385e5e0bea -> 84748bd36dff8cb77154623eac4c3f3647720072 ([`pr/gklog.49`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.49) -> [`pr/gklog.50`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.50), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.49-rebase..pr/gklog.50))<!-- end --> due to conflict with #35496
<!-- begin push-51 -->
Rebased 84748bd36dff8cb77154623eac4c3f3647720072 -> dbe85f39723568b54b02ee9fbce8585a865a6e11 ([`pr/gklog.50`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.50) -> [`pr/gklog.51`](https://github.com/ryanofsky/bitcoin/commits/pr/gklog.51), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gklog.50-rebase..pr/gklog.51))<!-- end --> due to conflicts with #35965

ryanofsky committed:
log refactor: Add support for custom log contexts

Custom log contexts allow overridding log formatting and adding metadata, such
as request ids or wallet names to log messages, while still using standard
macros for logging. This is used to replace WalletLogPrintf() functions with
LogInfo() calls in followup PR #30343.

ryanofsky committed:
kernel: Simplify logging API

Current kernel logging API is too complicated and too restrictive. This PR
tries to improves it.

I'd expect an API that supported logging to allow creating log streams with
different options and providing some way to specify which library operations
should be logged to which streams.

By contrast, the current API allows creating multiple log streams, but all the
streams receive the same messages because options can only be set globally and
the stream objects can't be passed to any kernel API functions. They are not
even referenced by the `btck_Context` struct which holds other shared state. If
no log streams are created, log messages are generated anyway, but they are
stored in a 1MB buffer and not sent anywhere, unless a log stream is created
later, at which point they are sent in bulk to that stream. More log streams
can be created after that, but they only receive new messages, not the buffered
ones. If log output is not needed, a btck_logging_disable() call is required to
prevent log messages from accumulating in the 1MB buffer. Calling this will
abort the program if any log streams were created before it was called, and
also abort the program if any new log streams are created later.

None of these behaviors seem necessary or ideal, and it would be better to
provide a simpler logging API that allows creating a log stream, setting
options on it, registering it with the `btck_Context` instance and receiving
log messages from it. Another advantage of this approach is that it could allow
(with #30342) different log streams to be used for different purposes, which
would be not be possible with the current interface.

stickies-v committed:
log test: verify log argument evaluation semantics

Test that LogInfo/LogWarning/LogError always evaluate their arguments
even when logging is disabled.

ajtowns pointed out this behavior was important and could affect non-logging
code if changed in
https://github.com/bitcoin/bitcoin/pull/34374#discussion_r2734793117

Co-authored-by: l0rinc <pap.lorinc@gmail.com>

stickies-v reviewed (commented):
I've finally started reviewing this PR in more detail, after my recent related work on kernel logging. I'm currently leaning Concept ACK, Approach NACK.

Concept ACK because:
- a large part of the bitcoinkernel interface is contextualized, e.g. `btck_ChainstateManager` is initialized with a `btck_Context`, so having contextualized logging seems like a natural fit
- bitcoinkernel is a library, we should be minimally opinionated on how it's used, so long as it's safe. If users have good use cases for e.g. multiple `ChainstateManager`'s, being able to separate log streams would be good

Approach NACK because I think the below downsides are bigger than the above upsides:
- logging interface becomes more cumbersome, i.e. caller has to figure out if a logger is in scope and if it should be used. It is easy to misuse (e.g. forget to log to context when it exists, which seems hard to enforce.)
- non-trivial code change / review cost, lots of merge conflicts, ...
 
I think this change might become more desirable if/when we have more demand for this feature (i.e. more actual use cases for contextualized logging) or when it becomes a less invasive code change to do so (e.g. if/when kernel is a separate codebase, or when it becomes feasible to use an interface like `m_log.info(...)`).

I've asked the kernel working group for use cases and motivations for contextualized logging, and so far I've not heard much (but I hope people will chime in here rather than me speaking on their behalf). If you have any, perhaps opening a tracking issue for that might be useful so we can better decide when this feature becomes worth it given the complexity? I'm definitely not against the idea itself, I just pragmatically think it's not worth the current cost.

stickies-v committed:
log test: add some test coverage on LogAcceptCategory

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

ryanofsky committed:
logging: Add LOG_REQUIRE_CONTEXT option

Add option to require context objects to be passed to log macros in a file or
block of code, so new logging calls can't forget to specify them and
accidentally log to the global logging stream.

ryanofsky committed:
log refactor: Drop Entry::should_ratelimit field

Drop Entry::should_ratelimit field in favor of Options::ratelimit field.

Dropping the Entry::should_ratelimit field makes buffered log messages
in m_msgs_before_open smaller, and avoids potential confusion because
the field is stored after rate limiting has been applied, so just gets
ignored. Conceptually it also makes sense to treat the ratelimit option
as an input to the logger rather than as an attribute of the message
being logged.

ryanofsky committed:
log refactor: log macro rewrite

Rewrite log macros to fix a number of issues: unnecessary strprintf
calls during fuzzing, confusing error messages when macros are called
with the wrong arguments, duplicated code with unexplained differences
and undocumented assumptions.

Since this is a rewrite, reading the new code in log.h first should be clearer
than starting from the diff.

Specific benefits of the new implementation are:

- Functionality is implemented once in a single `LOG_EMIT` macro instead of
  multiple times in different macros with diverging code paths and unexplained
  differences.

- Unnecessary `strprintf` calls are now skipped when logging is disabled (in
  bitcoind when `-noprinttoconsole -nodebuglogfile` options are used, and in
  tests and kernel applications when DisableLogging is called). This change
  should not affect bitcoind noticeably, but could speed up fuzz tests calling
  functions which log.

- Clearer error messages: If you pass a category to a macro which does not
  accept it, or forget to pass a category to a macro which requires it, you
  will see a direct message telling you to add or remove the category instead
  of expanded macro syntax errors.

- Previously it was possible to call `detail_LogWithSrcLoc` with
  inconsistent or invalid arguments, for example bypassing ratelimiting
  without it being explicit. New `LOG_EMIT` macro is compile-time safe
  and enforces the exact same restrictions as other macros.

- Previously "always evaluate arguments" behavior at Info and higher levels
  looked accidental and was undocumented
  (https://github.com/bitcoin/bitcoin/pull/34374#discussion_r2734793117). Now
  the behavior is documented and explicit.

- `NO_RATE_LIMIT` special case used to bypass rate limiting is dropped.
  The `LOG_EMIT` macro can control rate limiting (and other options if
  they are added in the future) while still using the same ratelimiting
  defaults as other macros.

Co-Authored-By: Hodlinator <172445034+hodlinator@users.noreply.github.com>
Co-Authored-By: stickies-v <stickies-v@protonmail.com>

ryanofsky committed:
Merge branch 'pr/relog' into pr/bclog

ryanofsky committed:
test-each-commit: Increase fetch depth

Needed due to base PR, can be dropped before merge

ryanofsky committed:
Merge branch 'pr/bclog' into pr/gklog

ryanofsky committed:
Merge branch 'pr/klog' into pr/gklog
